mitiq/folding_cirq.py

```diff
@@ -1,76 +1,234 @@
+                return folded
 
 
-def left_stretcher(circuit: Circuit, stretch: float) -> Circuit:
-    """Applies the map G -> G G^dag G to a subset of gates of the input circuit (first part of it).
-    Returns a circuit of depth approximately equal to stretch*len(circuit).
-    The stretch factor can be any real number within 1 and 3."""
+def fold_gates_at_random(circuit: Circuit, stretch: float, seed: Optional[int]) -> Circuit:
+    """Returns a folded circuit by applying the map G -> G G^dag G to a random subset of gates in the input circuit.
 
-    if not ((stretch >= 1) and (stretch <= 3)):
-        raise ValueError("The stretch factor must be a real number within 1 and 3.")
+    The folded circuit has a number of gates approximately equal to stretch * n where n is the number of gates in
+    the input circuit.
 
-    depth = len(circuit)
-    reduced_depth = int(depth * (stretch - 1) / 2)
-    sub_indices = list(range(reduced_depth))
-    return fold_gates(circuit, sub_indices)
+    Args:
+        circuit: Circuit to fold.
+        stretch: Factor to stretch the circuit by. Any real number in the interval [1, 3].
+        seed: [Optional] Integer seed for random number generator.
+
+    Note:
+        Folding a single gate adds two gates to the circuit, hence the maximum stretch factor is 3.
+    """
+    folded = deepcopy(circuit)
 
+    if np.isclose(stretch, 1.0, atol=1e-2):
+        return folded
 
-def fold_gates(circuit: Circuit, sub_indices: List[int]) -> Circuit:
-    """Applies the map G -> G G^dag G to a subset of gates of the input circuit
-    determined by sub_indices."""
+    if not 1 < stretch <= 3:
+        raise ValueError("The stretch factor must be a real number between 1 and 3.")
 
-    out = Circuit()
-    for j, gate in enumerate(circuit):
-        out += gate
-        if j in sub_indices:
-            out += inverse(gate)
-            out += gate
-    return out
+    if seed:
+        np.random.seed(seed)
 
+    ngates = len(list(folded.all_operations()))
+    num_to_fold = int(ngates * (stretch - 1.0) / 2.0)
 
-def local_folding(circuit: Circuit, stretch: float, stretcher: Callable=left_stretcher) -> Circuit:
-    """Applies the map G -> G G^dag G to a subset of gates of the input circuit.
-    Returns a circuit of depth approximately equal to stretch*len(circuit).
-    The stretch factor can be any real number >= 1."""
+    for _ in range(num_to_fold):
+        # TODO: This allows for gates to be folded more than once, and folded gates to be folded.
+        #  Should this be allowed?
+        moment_index = np.random.choice(len(folded))
+        gate_index = np.random.choice(len(folded[moment_index]))
+        # TODO: It could be expensive to call fold_gate...(...) which makes a deepcopy of the circuit each call.
+        #  ==> Possible fix: Have fold_gate_at_index_in_moment(...) modify the circuit in place.
+        folded = fold_gate_at_index_in_moment(folded, moment_index, gate_index)
```

Comment:
Any reason not to change this and remove the todo?
```suggestion
        folded = _fold_gate_at_index_in_moment(folded, moment_index, gate_index)
```

Reply:
Likewise missed this, thanks!